AIChat refactor to support standalone and persistent conversations

petemill opened the pull request:
This PR refactors the ai_chat component on the browser-side and in the UI to allow for
- Multiple conversations
- Conversations to be independent of content (WebContents/Tab)

This is achieved with the following architecture (more details at https://github.com/brave/brave-browser/issues/31623):
- AIChatService owns all the conversation metadata
- ConversationHandler owns the data for conversations (chat messages, suggested questions) and performs all the operations via the EngineConsumer (which is unchanged).
- Content (title, url, page content) is attached to a conversation via an AssociatedContentDelegate. This is implemented by AssociatedContentDriver (and derives AIChatTabHelper for WebUI and an iOS equivalent). When a conversation is related to some content, one of these is attached. Standalone conversations do not have associated content and therefore this delegate is null.

AIChatService -> ConversationHandler -> AssociatedContentDelegate [AssociatedContentDriver [AIChatTabHelper]]

- Conversations which had associated content, but the content is not live anymore (i.e. the Tab was navigated away or closed), receive an AssociatedArchiveContent, which keeps the url and last title and page content so that the conversation can be continued.

Future TODO (other PRs)
- [ ] Connect the storage from https://github.com/brave/brave-core/pull/19503
- [ ] Connect the UI from https://github.com/brave/brave-core/pull/24633
- [ ] iOS "other conversations" UI
- [ ] android "other conversations" UI

Resolves https://github.com/brave/brave-browser/issues/31623

<!-- CI-related labels that can be applied to this PR:
* CI/run-audit-deps (1) - check for known npm/cargo vulnerabilities (audit_deps)
* CI/run-network-audit (1) - run network-audit
* CI/run-upstream-tests - run Chromium unit and browser tests on Linux and Windows (otherwise only on Linux)
* CI/run-linux-arm64, CI/run-macos-x64, CI/run-windows-arm64, CI/run-windows-x86 - run builds that would otherwise be skipped
* CI/skip - do not run CI builds (except noplatform)
* CI/skip-linux-x64, CI/skip-android, CI/skip-macos-arm64, CI/skip-ios, CI/skip-windows-x64 - skip CI builds for specific platforms
* CI/skip-upstream-tests - do not run Chromium unit, or browser tests (otherwise only on Linux)
* CI/skip-all-linters - do not run presubmit and lint checks
* CI/storybook-url (1) - deploy storybook and provide a unique URL for each build

(1) applied automatically when some files are changed (see: https://github.com/brave/brave-core/blob/master/.github/labeler.yml)
-->

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

darkdh commented on browser/ui/ai_chat/print_preview_extractor.cc:
```diff
@@ -166,342 +167,383 @@ class PreviewPageTextExtractor {
-}
+  void CreatePrintPreview() {
+    if (profile_->GetPrefs()->GetBoolean(::prefs::kPrintPreviewDisabled)) {
+      DeleteSelfAndSendResult("");
+      return;
+    }
+    content::RenderFrameHost* rfh = GetRenderFrameHostToUse(web_contents_);
+    if (rfh) {
+      if (!print_render_frame_.is_bound()) {
+        rfh->GetRemoteAssociatedInterfaces()->GetInterface(
+            &print_render_frame_);
```
Why are we changing this class to be self deleted? `PrintPreviewExtractor` was owned by AIChatUI to ensure it won't outlive the chat panel and also ensure there is only one instance exists at a time per renderer because `mojo::AssociatedRemote<printing::mojom::PrintRenderFrame> print_render_frame_` can only be bound to one receiver at the same time. Self deleting will break that guarantee.

bridiver replied:
That isn't the only problem here, it could also outlive the webcontents

bridiver replied:
the webcontents issue could be fixed by passing the frame tree node id and using it to get the webcontents each time, but there may be other issues as @darkdh mentioned

bridiver replied:
It also seems very easy to accidentally leak this class if there's some code path that is missing the call to delete itself.

petemill replied:
>  was owned by AIChatUI to ensure it won't outlive the chat panel 

As explained, Conversations are now decoupled from `AIChatUI`. You can close a conversation in a panel, mid-generation, and pick up on it in the full-page, or sometime later. Any `AIChatUI` can bind to any conversation from any `WebContents` (or standalone conversations but that's not relevant here). In fact, `AIChatUI` doesn't know which conversation is open in the UI, the renderer binds to the `ConversationHandler` directly. Not to mention the previous pattern was very confusing IMO since it used an observer, but we want to guarantee that only 1 observer is attached (which it wasn't). It is a 1:1 relationship purely so that we can run code in `chrome/browser/ui` so some kind of delegate or provider pattern makes more sense.

> ensure there is only one instance exists at a time per renderer because mojo::AssociatedRemote<printing::mojom::PrintRenderFrame> print_render_frame_ can only be bound to one receiver at the same time

What guarantee was there previously in `PrintPreviewExtractor` that isn't there now? In this PR, `ConversationDriver::GetContent` is guarded by `is_page_text_fetch_in_progress_` and that's the only place which calls `AIChatTabHelper::GetPageContent` which is the only place which runs the print preview extractor. All page content fetching is guaranteed to only happen once.

> it could also outlive the webcontents

Yes that's an issue. This pattern is also used in `FetchPageContent`, and in Chromium's `FetchRssLinks`, which is admittedly simpler and uses a mojom disconnect handler to `SendResultAndDelete` when the `WebContents` is destroyed. In this case, we could abandon when the `WebContents` is destroyed because the callback to the `AIChatTabHelper` wouldn't be valid anymore anyway. I suppose I'll change this to be yet another delegate. It has to be a delegate and not directly owned because `AIChatTabHelper` is in `components/ai_chat/content/browser`, but `PrintPreviewExtractor` requires access to `chrome/browser` and `chrome/browser/ui`. It seems less convenient to move `AIChatTabHelper` to `brave/browser/ui`.

darkdh replied:
> What guarantee was there previously in PrintPreviewExtractor that isn't there now? 
```  
print_preview_extractor_ = std::make_unique<PrintPreviewExtractor>(
      active_chat_tab_helper_->web_contents(), profile_, is_pdf);
```
To make sure there is only one instance per web_contents disregarding previous one is still running or not.

petemill replied:
This is addressed now. AIChatTabHelper is provided with an `AIChatTabHelper::PrintPreviewExtractionDelegate` instance, `PrintPreviewExtractor`, which itself owns an instance of `PrintPreviewExtractorInternal`, which it creates and resets on each request to `Extract()`, to match the current implementation of construction for each extraction request.

darkdh replied:
The new delegate approach is neater than the self deleted one and even original subclassing `printing::mojom::PrintPreviewUI` in Leo panel UI.

darkdh commented on browser/ui/ai_chat/print_preview_extractor.cc:
```diff
@@ -166,342 +167,383 @@ class PreviewPageTextExtractor {
-      std::move(pdf_region.region),
-      base::BindOnce(&PrintPreviewExtractor::OnGetOCRResult,
-                     weak_ptr_factory_.GetWeakPtr()),
-      active_chat_tab_helper_->GetMaxPageContentLength(),
```
GetMaxPageContentLength was meant for saving unnecessary OCR computing power when we already exceeds the context limit and will be truncated. Maximum 20 pages doesn't prevent this from happening. (It's possible that we reach model context limit first before reaching 20 pages) It is used to prevent edge case like multiple blank pages or small amount of context per page which will take very long to finish OCR when there are significant amount of pages.

petemill replied:
Is 20 significant? Do you have a different proposal to solve the issue?

As explained over DM a few days ago:
> Also removed max page content consideration since:
> 1) Model can be changed after fetching the content
> 2) Multiple callbacks could occur with different page content limits from different conversations.
> 3) We need to know the (reasonable) total content length to report the percentage of content that Leo has been sent.
> I did consider sending the requested page content maximum length for each GetContent call, but that does not solve all the issues. Since there is a maximum limit of 20 print preview pages, it seems it's ok.

Please remember that multiple conversations can be attached to the same Tab content, and can have different models selected. One other way I could pass a max in is getting the max of all configured models. But that doesn't solve 3). The user would think that Leo is using (almost) 100% the page content when it could be much lower (although still limited to 20 pages).

darkdh replied:
20 is significant when running OCR. If we care so much about reporting percentage of the page content Leo uses, then we can drop max length within PrintPreviewExtrator as you do now. And we should create a follow-up issue for `PrintPreviewExtrator` to report total page count when it hit 20 pages limits so users would know Leo only use parts of the page content.

bridiver commented on browser/ai_chat/ai_chat_service_factory.cc:
```diff
@@ -0,0 +1,88 @@
+std::unique_ptr<KeyedService>
+AIChatServiceFactory::BuildServiceInstanceForBrowserContext(
+    content::BrowserContext* context) const {
+  auto skus_service_getter = base::BindRepeating(
```
whys is this passing a getter instead of the skus service itself?

petemill replied:
This code is just moved. We should fix that, but not here.

bridiver replied:
can we create follow-up issues to track this stuff?

bridiver commented on browser/ai_chat/ai_chat_service_factory.cc:
```diff
@@ -0,0 +1,88 @@
+
+content::BrowserContext* AIChatServiceFactory::GetBrowserContextToUse(
+    content::BrowserContext* context) const {
+  if (!Profile::FromBrowserContext(context)->IsRegularProfile() ||
```
use `ProfileSelections::Builder().WithRegular(ProfileSelection::kOwnInstance)`

petemill replied:
I just merged another PR which adds https://github.com/brave/brave-core/blob/master/browser/ai_chat/ai_chat_utils.cc#L15-L20:

```
bool IsAllowedForContext(content::BrowserContext* context,
                         bool check_policy /*=true*/) {
  Profile* profile = Profile::FromBrowserContext(context);
  return profile->IsRegularProfile() && ai_chat::features::IsAIChatEnabled() &&
         (!check_policy || IsAIChatEnabled(profile->GetPrefs()));
}
```

That's fine to stay that way? I was going to update here to use this new util, but we'd rather use the `ProfileSelections` path instead?

petemill replied:
shouldn't this be `.WithRegular(ProfileSelection::kOriginalOnly)` to make sure the service isn't built for incognito? Don't we in-fact want the default behavior of `ProfileKeyedServiceFactory`?
> Example of a service that does not exist in OTR (default behavior)

bridiver commented on browser/ai_chat/ai_chat_service_factory.cc:
```diff
@@ -0,0 +1,88 @@
+  return std::make_unique<AIChatService>(
+      ModelServiceFactory::GetForBrowserContext(context),
+      std::move(credential_manager), user_prefs::UserPrefs::Get(context),
+      (g_brave_browser_process->process_misc_metrics())
```
why the parens around this? @DJAndries have you had a change to do anything with this yet to make it less clunky to access?

petemill replied:
Because of the ternary? formatter isn't complaining, but happy to change...

bridiver replied:
shouldn't be needed here

bridiver commented on browser/ai_chat/ai_chat_service_factory.cc:
```diff
@@ -0,0 +1,88 @@
+          : nullptr,
+      context->GetDefaultStoragePartition()
+          ->GetURLLoaderFactoryForBrowserProcess(),
+      std::string(version_info::GetChannelString(chrome::GetChannel())));
```
why are you creating s string from this? You should pass it as string_view and then store it as a string in AIChatService

petemill replied:
This PR moves a lot of existing code from ConversationDriver to the new `AIChatService`, `ConversationHandler`, and `AssociatedContentDriver`. I'd rather not increase the scope and fix unrelated issues, unless it's just a couple, as the merge will be painful if other changes are merged to master in the meantime. It'd be great to focus the review on the new ownership and binding patterns, ahead of the introduction of the SQL storage that is pending in another branch.

nullhook commented on components/ai_chat/core/browser/ai_chat_service.cc:
```diff
@@ -0,0 +1,441 @@
+    // and still in memory.
+    conversation = GetConversation(conversation_uuid);
+  }
+  if (!conversation) {
+    // New conversation needed
+    conversation = CreateConversation();
```
should this case called `CreateConversationHandlerForPageContent`? i see the logic is the same

yrliou commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -1056,6 +1066,65 @@ void ConversationHandler::MaybeSeedOrClearSuggestions() {
+void ConversationHandler::OnGetStagedEntriesFromContent(
+    const std::optional<SearchQuerySummary>& search_query_summary) {
+  // Check if all requirements are still met.
+  if (!search_query_summary || !chat_history_.empty() ||
```
Should we also check user opt-in here?

petemill replied:
It's being checked in `MaybeFetchOrClearContentStagedConversation` and there's no way currently to opt-out whilst that's happening, but I'll add the check anyway.

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/aichat-services/macos/arm64/storybook.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/aichat-services/linux/x64/storybook.html)

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/aichat-services/windows/x64/storybook.html)

petemill commented:
I've marked this as ready to review since IMO it's feature complete - maintains all existing functionality whilst allowing for full-page and conversation storage (contained on other PRs). I'd love to get the C++ and TS reviewed ASAP to avoid the constant rebase game. Tests are passing, but iOS refactor is still not complete.
Please note that a lot of code has been moved and not everything that Github shows as new _is_ new. I've fixed up a fair amount of code style issues, but not everything as I think we're at the limit of scope for this PR. Having said that, I've added some tests for things that were previously untested, and tested most new algorithms.

petemill commented:
I've removed the change from Untrusted WebUI to Trusted WebUI from this PR so that we can focus the security review for that on a smaller PR with other changes that need security review (the full-page UI PR).

petemill commented on browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc:
```diff
@@ -101,121 +84,22 @@ AIChatUIPageHandler::AIChatUIPageHandler(
+  // TODO(petemill): figure out why java needs access to conversation context
+  // web contents and this can't be done in UI WebContents only. Taget
+  // WebContents
   auto* context_web_contents = (active_chat_tab_helper_)
```
@SergeyZhukovsky can this just always be the WebUI's `web_contents()`?

SergeyZhukovsky replied:
it used to pass a Leo query after voice recognition work https://github.com/brave/brave-core/blob/master/android/java/org/chromium/chrome/browser/brave_leo/BraveLeoVoiceRecognitionHandler.java#L69.

petemill replied:
ah ok thanks. In this refactor, a Conversation might not have a WebContents context, since it might be standalone. I think we could change this to pass a simple string conversation id?

petemill replied:
@SergeyZhukovsky I have pushed an android change to target a specific conversation when calling `handleVoiceRecognition`. Please take a look 7e81dc4b1e4a6d38f4152b348c051ccda49e34dd

petemill commented on browser/ui/ai_chat/print_preview_extractor.cc:
```diff
@@ -0,0 +1,562 @@
+    pdf_to_bitmap_converter_->GetPdfPageCount(
+        pdf_region_.Duplicate(),
+        base::BindOnce(&PreviewPageTextExtractor::OnGetPageCount,
+                       base::Unretained(this)));
```
already existing via @darkdh

petemill commented on browser/ui/ai_chat/print_preview_extractor.cc:
```diff
@@ -0,0 +1,562 @@
+      SendResult("");
+      return;
+    }
+    memcpy(pdf_region.mapping.memory(), data->data(), data->size());
```
already existing via @darkdh

petemill commented on browser/ui/ai_chat/print_preview_extractor.cc:
```diff
@@ -0,0 +1,562 @@
+      pdf_to_bitmap_converter_->GetBitmap(
+          pdf_region_.Duplicate(), current_page_index_,
+          base::BindOnce(&PreviewPageTextExtractor::OnGetBitmap,
+                         base::Unretained(this)));
```
already existing via @darkdh

petemill commented on browser/ui/ai_chat/print_preview_extractor.cc:
```diff
@@ -0,0 +1,562 @@
+        pdf_to_bitmap_converter_.BindNewPipeAndPassReceiver());
+    pdf_to_bitmap_converter_.set_disconnect_handler(
+        base::BindOnce(&PreviewPageTextExtractor::BitmapConverterDisconnected,
+                       base::Unretained(this)));
```
already existing via @darkdh

petemill commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -112,9 +116,9 @@ class PageContentFetcher {
     // after it is destroyed.
     content_extractor_.set_disconnect_handler(base::BindOnce(
-        &PageContentFetcher::DeleteSelf, base::Unretained(this)));
+        &PageContentFetcherInternal::DeleteSelf, base::Unretained(this)));
```
Just renamed the class

petemill commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -112,9 +116,9 @@
+        &PageContentFetcherInternal::DeleteSelf, base::Unretained(this)));
     content_extractor_->ExtractPageContent(base::BindOnce(
-        &PageContentFetcher::OnTabContentResult, base::Unretained(this),
+        &PageContentFetcherInternal::OnTabContentResult, base::Unretained(this),
```
Just renamed the class

petemill commented on components/ai_chat/content/browser/page_content_fetcher.cc:
```diff
@@ -127,7 +131,7 @@
     }
     content_extractor_.set_disconnect_handler(base::BindOnce(
-        &PageContentFetcher::DeleteSelf, base::Unretained(this)));
+        &PageContentFetcherInternal::DeleteSelf, base::Unretained(this)));
```
Just renamed the class

petemill commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -0,0 +1,1410 @@
+                       std::move(navigation_id)));
+  };
+  GeneratePageContent(base::BindOnce(std::move(on_content_retrieved),
+                                     base::Unretained(this),
```
changed

petemill commented on components/ai_chat/resources/page/ai_chat_ui.html:
```diff
@@ -4,12 +4,12 @@
+    <link rel="stylesheet" href="//resources/brave/css/reset.css">
+    <link rel="stylesheet" href="//resources/css/text_defaults.css">
+    <link rel="stylesheet" href="//resources/brave/fonts/manrope.css">
+    <link rel="stylesheet" href="//resources/brave/fonts/poppins.css">
```
Not applicable for WebUI

fallaciousreasoning commented on browser/ui/ai_chat/ai_chat_tab_helper_unittest.cc:
```diff
@@ -0,0 +1,220 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/ai_chat/content/browser/ai_chat_tab_helper.h"
```
nit but should this file live next to `ai_chat_tab_helper.cc`?

petemill replied:
Two reasons:
1. I needed to use `TestingProfile` and `RenderViewHostTestHarness` to get the TabHelper, ContentFaviconDriver and FaviconService setup without complaining or crashing. Both those are in `chrome/` so can't be called from `components/`.

2. I thought I was using the `PrintPreviewExtractor` implementation in the unit test. Admittedly they should be decoupled to 2 unit tests. And it looks like I'm not actually using it, the implementation is covered by the general AI Chat browser tests. So I guess I can at least move the test to `brave/browser/ai_chat/`.

fallaciousreasoning commented on components/ai_chat/core/browser/associated_content_driver.h:
```diff
@@ -0,0 +1,167 @@
+  // Handlers that are interested in this content for the current navigation.
+  std::set<raw_ptr<ConversationHandler>> associated_conversations_;
+
+  // Store the unique ID for each "page" so that
```
This is just for the current page, right?
```suggestion
  // Store the unique ID for the current "page" so that
```

petemill replied:
Yes, thanks - this comment was copied from the original `ConversationDriver` which handled multiple pages, although I guess your wording would have been better then too

fallaciousreasoning commented on components/ai_chat/core/browser/associated_content_driver.h:
```diff
@@ -0,0 +1,167 @@
+
+  base::ObserverList<Observer> observers_;
+  std::unique_ptr<base::OneShotEvent> on_page_text_fetch_complete_;
+  bool is_page_text_fetch_in_progress_ = false;
```
I think rather than having two separate variables here we could only set `on_page_text_fetch_complete_` while we're fetching - then the fetch in progress check becomes

```cpp
if (on_page_text_fetch_complete_)
```

petemill replied:
I did that - thanks for the suggestion b881722559bc5a51dd097570d9f1bd0522e2a4e2

fallaciousreasoning commented on components/ai_chat/core/browser/associated_archive_content.h:
```diff
@@ -0,0 +1,51 @@
+
+namespace ai_chat {
+
+class AssociatedArchiveContent
```
A comment about exactly what this is for would be handy :smile:

petemill replied:
Added 8c401ad7b3bf8f2df987571698d4e0e4fc9256d2

fallaciousreasoning reviewed (commented):
The frontend stuff looks good to me - I'm going to take a bit more time to digest everything else :laughing: (there's a lot)

petemill commented on components/ai_chat/content/browser/ai_chat_tab_helper.h:
```diff
@@ -44,37 +37,69 @@ class AIChatTabHelper : public content::WebContentsObserver,
+
+    virtual ~PrintPreviewExtractionDelegate() = default;
+    // Get the current text from the WebContents using Print Preview. This
+    // should not be called
```
er, maybe I got distracted mid-sentence! I think maybe I meant this should not be called multiple times?! I don't know. Removed that part.

petemill commented on components/ai_chat/content/browser/ai_chat_tab_helper.h:
```diff
@@ -137,12 +169,19 @@ class AIChatTabHelper : public content::WebContentsObserver,
   raw_ptr<content::WebContents> inner_web_contents_ = nullptr;
+
+  // TODO(petemill): Use signal to allow for multiple callbacks
+  ConversationHandler::GetPageContentCallback
```
It doesn't need to IMO because that is actually handled at the higher level (`AssociatedContentDriver`). However, this class doesn't know that. For non-pdf page's, `GetPageContent` can handle multiple calls at the same time (though it doesn't queue them, they'll just run independently). But for PDF documents, the existing `pending_get_page_content_callback_` will get called with no content and only the latest call will be used. But it's not an issue with the way `GetPageContent` is called right now. So, the TODO is there as a direction in case `AIChatTabHelper::GetPageContent` gets used in other ways in the future.

petemill commented on components/ai_chat/content/browser/ai_chat_tab_helper.cc:
```diff
@@ -168,24 +135,25 @@ void AIChatTabHelper::WebContentsDestroyed() {
-  is_same_document_navigation_ = navigation_handle->IsSameDocument();
-  pending_navigation_id_ = navigation_handle->GetNavigationId();
-
+  is_same_document_navigation_ = load_details.is_same_document;
```
> also, fwiw I'm a little skeptical of the title change heuristic, but that's not for this PR 😆
It might not be perfect but I'd rather false negatives than false positives.

Although I did just notice a problem when testing - back navigations don't call `TitleWasSet`, they just update the title. The title is in `NavigationEntryCommitted`'s params though, so I can check if it's different there too. I'll have to add a test for this in the meatier history implementation PR.

> presumably the same document navigations will get a different id 
That's a good point. If I were to update the reference navigation ID without calling OnNewPage, same-document navigations which never result in a new conversation could get some bugs where results of async operations will be ignored because the navigation ID changed. And also the conversation would lose it's associated with the original navigation. Without doing that though, the conversation won't ever be associated with the current navigation when the user navigates back to it.
Seems like ideally we'd want it to be associated with multiple navigations. Should be simple enough as `AIChatService`'s map of navigation ID to conversation ID will be fine with the same conversation appearing multiple times.

petemill replied:
I have a "fix" for this on a separate branch - bf1781c5194077129c1fcbecdf7e75cf9d4a514a. That fix makes it so that the additional navigation IDs will be associated with the same conversation which will fix your scenario. I'm going to think more about it, and make a test, before I put it on this branch. It's not too neccessary for this branch since the conversation preservation and re-joining is behind a feature flag.

fallaciousreasoning replied:
Yup, happy to leave this for a followup PR

petemill commented on components/ai_chat/content/browser/ai_chat_tab_helper.cc:
```diff
@@ -265,16 +232,18 @@ void AIChatTabHelper::OnInterceptedPageContentChanged() {
+  bool is_pdf = IsPdf(web_contents());
+  if (is_pdf && !is_pdf_a11y_info_loaded_) {
     if (pending_get_page_content_callback_) {
+      // TODO(petemill): Queue the callback in a Signal, instead of only
```
Same TODO as above, really, although this one more specific to a functional problem if called multiple times, instead of an optimization issue at the above TODO's location.

> It doesn't need to IMO because that is actually handled at the higher level (AssociatedContentDriver). However, this class doesn't know that. For non-pdf page's, GetPageContent can handle multiple calls at the same time (though it doesn't queue them, they'll just run independently). But for PDF documents, the existing pending_get_page_content_callback_ will get called with no content and only the latest call will be used. But it's not an issue with the way GetPageContent is called right now. So, the TODO is there as a direction in case AIChatTabHelper::GetPageContent gets used in other ways in the future.

...but if you mean should I have said "OneShotEvent" and not "Signal", because `Signal()` is what you call on a `OneShotEvent` - then yes! 😄

petemill commented on components/ai_chat/content/browser/ai_chat_tab_helper.cc:
```diff
@@ -300,20 +269,47 @@ void AIChatTabHelper::GetPageContent(GetPageContentCallback callback,
+void AIChatTabHelper::OnExtractPrintPreviewContentComplete(
+    ConversationHandler::GetPageContentCallback callback,
+    std::string content) {
+  // Invalidation token not applicable for print preview OCR
```
The third parameter of `GetPageContentCallback` is the `invalidation_token`, which helps not refetch adjacent content (e.g. video transcripts) if the underlying content didn't change between calls to `GetPageContent` (which is called every time the user submits a chat message). On some (most!) sites we want to re-fetch every time the content because it might have updated and it's just a simple local-only content extraction. However on  other sites, the content fetch results in network calls or might be expensive otherwise _and_ we know the content is static for now, so we use an invalidation token to say "use your cache for this content if the token is still the same as the one we extract cheaply".

fallaciousreasoning replied:
sweet, sgtm

petemill commented on components/ai_chat/core/common/mojom/ai_chat.mojom:
```diff
@@ -226,69 +227,155 @@ struct ActionGroup {
+
+  // TODO(petemill): WebUI clients won't need this when they can
+  // request via FaviconSource URLs via trusted webui.
+  GetFaviconImageData(string conversation_id) => (
```
Yeah the comment is aiming to convey that we can't use `chrome://favicon[2]` because it's explicitly not available in Untrusted WebUI for security reasons.

petemill commented on components/ai_chat/core/common/mojom/ai_chat.mojom:
```diff
@@ -226,69 +227,155 @@ struct ActionGroup {
+
+// Browser-side handler for a Conversation
+interface ConversationHandler {
+  // TODO(petemill): GetInitialState => (InitialState conversation_state);
```
Yeah it's still a TODO for optimization. Right now we're calling multiple functions when initially binding a conversation in the WebUI when it's clear that a UI will need all the initial data.

darkdh commented on components/ai_chat/core/browser/conversation_handler_unittest.cc:
```diff
@@ -1091,9 +1091,8 @@ TEST_P(PageContentRefineTest, TextEmbedderInitialized) {
-    SCOPED_TRACE(testing::Message()
-                 << "IsInitialized: " << test_case.is_initialized
-                 << ", Initialize result: " << test_case.initialize_result);
+    LOG(ERROR) << "IsInitialized: " << test_case.is_initialized
+               << ", Initialize result: " << test_case.initialize_result;
```
Forgot to remove these?

petemill commented on browser/ui/ai_chat/print_preview_extractor.cc:
```diff
@@ -0,0 +1,562 @@
+        printing::GetFrameToPrint(web_contents_.get());
+    if (!render_frame_host) {
+      DLOG(ERROR) << "No render frame host for print preview";
+      SendResult("");
```
Well in current implementation it also early returns but yes it doesn't send the result. I now see that this is per-page. What do you recommend? Is this not fatal if it happens on every page? Could it happen?

darkdh replied:
Early return doesn't mean we can sending result. just remove the sending result back.
This is unlikely to happen. Even it happens on all pages, we should handle it in `MetafileReadyForPrinting` which is called when `PrintRenderFrameHelper::FinalizePrintReadyDocument` finishes

darkdh replied:
The effect of early return here is we just don't send the page into compositor then it would match `printing::mojom::DidPreviewDocumentParamsPtr->expected_pages_count` when one or more pages went wrong when calling `FinishDocumentComposition`. So nothing is fatal here.

darkdh replied:
This check was originally extracted from `//chrome/browser/ui/webui/print_preview/print_preview_ui.cc` but the difference would be we don't have a print preview dialog which they check its rfh here so it is possible for `print_preview_ui.cc` to have empty rfh when dialog closed. But for our use case that won't happen, the only way it can be nullptr is tab closed.

petemill replied:
So early return is good enough and the original callback will still eventually get fired? 8c401ad7b3bf8f2df987571698d4e0e4fc9256d2

darkdh replied:
yes

petemill commented on browser/ui/ai_chat/ai_chat_tab_helper_unittest.cc:
```diff
@@ -0,0 +1,220 @@
+  bool is_print_preview_supported_ = true;
+};
+
+class AIChatTabHelperUnitTest_NoPrintPreview : public AIChatTabHelperUnitTest {
```
On master you didn't use a test suite, we have a single test with all the variations https://github.com/brave/brave-core/blob/60cad78225b8727ae291ca1f7cbfde417557c2d7/components/ai_chat/core/browser/conversation_driver_unittest.cc#L448. I created this extra test class just so we could add some tests to verify some functionality with print preview disabled.

And I think it's debatable since not all print preview tests apply to print preview being disabled, and having conditionals on expected functions and expected results adds complexity inside test bodies that can lead to test bugs. 

But I've been thinking about it and I do understand the preference to see the variation in a single test, and see value running all tests with all combinations of params. So I've pushed that change. b234336d95a7fad94b3e55af3af0c087f74e9a8f

petemill commented on components/ai_chat/content/browser/ai_chat_tab_helper.cc:
```diff
@@ -161,24 +164,25 @@ void AIChatTabHelper::WebContentsDestroyed() {
-
+  // UniqueID will provide a consitent value for the entry when navigating
+  // through history, allowing us to re-join conversations and navigations.
+  int pending_navigation_id = load_details.entry->GetUniqueID();
```
That's right

petemill commented on components/ai_chat/core/common/mojom/ai_chat.mojom:
```diff
@@ -226,9 +226,79 @@ struct ActionGroup {
+
+  // Receive specified Conversation
+  BindConversation(
+      string conversation_id,
```
All references to conversation id have been changed to conversation uuid 8c401ad7b3bf8f2df987571698d4e0e4fc9256d2

petemill commented on components/ai_chat/resources/page/chat_ui.tsx:
```diff
@@ -14,26 +14,91 @@ import '@brave/leo/tokens/css/variables.css'
+    // Observe when default conversation changes
+    const onNewDefaultConversationListenerId =
+      getAPI().UIObserver.onNewDefaultConversation.addListener(() => {
+        setDefaultConversationToken(new Date().getTime())
```
I'll admit this part is more complicated than I'd like, even though it's working well. I'll have another look at it all tomorrow.

petemill commented on components/ai_chat/content/browser/ai_chat_tab_helper.cc:
```diff
@@ -300,20 +269,47 @@ void AIChatTabHelper::GetPageContent(GetPageContentCallback callback,
+      print_preview_extraction_delegate_ != nullptr) {
+    // Get content using a printing / OCR mechanism, instead of
+    // directly from the source.
+    print_preview_extraction_delegate_->Extract(
```
fixed by removing that part of the comment!

darkdh commented on components/ai_chat/core/browser/ai_chat_service.cc:
```diff
@@ -0,0 +1,373 @@
+    // not a real / persistable conversation
+    auto uuid = conversation_handler->get_conversation_uuid();
+    conversation_observations_.RemoveObservation(conversation_handler);
+    conversation_handlers_.erase(uuid);
```
Not sure if we should also update `content_conversations_`, it can be replaced when a failed lookup happend in `GetOrCreateConversationHandlerForContent` but if it doesn't get accessed for a long time or even content no longer exists, the entry would stay in memory forever until shutdown.

petemill replied:
Yes we can erase any entries in the NavigationID to ConversationID map here. But in general there is a problem with that map (behind the AIChatConversationHistory feature flag) in that when the content for a conversation is navigated away we don't know how long to keep the entry in that map for in order to support matching the conversation on back-navigation. I think it's probably ok to keep it until the Tab closes, whilst the navigation ID is alive somewhere in the original WebContents navigation entry history because I think when a Tab is re-opened, its history items get different IDs than the original WebContents. And it's not much data.

petemill replied:
For now, I've added erasing `content_conversations_` based on values when a conversation is erased. We'll have to handle when content is erased in https://github.com/brave/brave-core/pull/25339

yrliou commented on components/ai_chat/core/browser/ai_chat_service.cc:
```diff
@@ -0,0 +1,373 @@
+    // and still in memory.
+    conversation = GetConversation(conversation_uuid);
+  }
+  if (!conversation) {
+    // New conversation needed
+    conversation = CreateConversation();
+  }
+  // Provide the content delegate, if allowed
+  if (associated_content &&
+      base::Contains(kAllowedSchemes, associated_content->GetURL().scheme())) {
+    conversation->SetAssociatedContentDelegate(associated_content);
+  }
+  // Record that this is the latest conversation for this content. Even
+  // if we don't call SetAssociatedContentDelegate, the conversation still
+  // has a default Tab's navigation for which is is associated. The Conversation
+  // won't use that Tab's Page for context.
+  content_conversations_.insert_or_assign(
+      associated_content_id, conversation->get_conversation_uuid());
+
+  return conversation;
```
These seems are duplicate codes with CreateConversationHandlerForContent.
Maybe just `return conversation ? conversation : CreateConversationHandlerForContent(associated_content_id, associated_content);`?

petemill replied:
De-duplicated with `MaybeAssociateContentWithConversation`

yrliou commented on components/ai_chat/core/browser/ai_chat_service.cc:
```diff
@@ -0,0 +1,373 @@
+
+void AIChatService::MaybeEraseConversation(
+    ConversationHandler* conversation_handler) {
+  if (!conversation_handler->IsAnyClientConnected() &&
+      (!features::IsAIChatHistoryEnabled() ||
+       !conversation_handler->HasAnyHistory())) {
```
Maybe want to include the case where all entries in history was staged entries?

petemill replied:
done c4bf8886b04

yrliou commented on components/ai_chat/core/browser/associated_content_driver.cc:
```diff
@@ -0,0 +1,284 @@
+#include "brave/components/ai_chat/core/browser/utils.h"
+#include "net/base/url_util.h"
+#include "net/traffic_annotation/network_traffic_annotation.h"
+namespace ai_chat {
```
nit: newline before this plz

yrliou commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -0,0 +1,1385 @@
+  receivers_.Add(this, std::move(receiver));
+  conversation_ui_handlers_.Add(std::move(conversation_ui_handler));
+  OnClientConnectionChanged();
+  // In some cases, this page handler hasn't been created and remote might not
+  // have been set yet.
+  // ex. A user may ask a question from the location bar
+  if (!pending_conversation_entry_.is_null()) {
+    OnHistoryUpdate();
+  }
```
I don't quite understand this, what's the relationship between whether pending_conversation_entry is null and if page handler is there and remote has been set?
Also, why is OnHistoryUpdate called when binding conversation?

petemill replied:
This was copied from the existing client connection code: https://github.com/brave/brave-core/blob/494f3a12db100e8c90c116be48ea08d47990a177/browser/ui/webui/ai_chat/ai_chat_ui_page_handler.cc#L108-L114

Doesn't make sense to me either - will remove and see if there's an issue with pending entry from the omnibox

yrliou reviewed (commented):
Haven't finished going through conversation_handler.cc/h and tests, will continue look into it.

petemill commented:
I've finished refactoring the iOS parts and it is now building locally (at least the C++ / obj-C parts). I've changed the swift code but had to move some things around and make some things async, and since it's pretty much my first time with swift please review it thoroughly @Brandon-T and make any changes you feel are neccessary. There's a better refactor there where we pass the mojom receivers and remotes directly to the swift class and stop proxying everything, but it's probably better to make that change as a follow-up, and keep the architecture roughly the same here.

Brandon-T commented on ios/browser/api/ai_chat/ai_chat.mm:
```diff
@@ -72,82 +69,60 @@ - (void)dealloc {
+          std::move(current_content_)));
+}
+
+- (void)createNewConversation {
```
If this function is never called in Swift, so we can remove it from the header (it doesn't need to be exported).

petemill replied:
It is called from swift: https://github.com/brave/brave-core/blob/003ef36fa17d7693d9e2dcd07bdfccfb1266be95/ios/brave-ios/Sources/AIChat/ModelView/AIChatViewModel.swift#L176

yrliou commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -0,0 +1,1408 @@
+        base::BindRepeating(
+            [](GeneratedTextCallback received_callback,
+               mojom::ConversationEntryEventPtr rewrite_event) {
+              constexpr char kResponseTagPattern[] =
+                  "<\\/?(response|respons|respon|respo|resp|res|re|r)?$";
+              if (!rewrite_event->is_completion_event()) {
+                return;
+              }
+
+              std::string suggestion =
+                  rewrite_event->get_completion_event()->completion;
+
+              base::TrimWhitespaceASCII(suggestion, base::TRIM_ALL,
+                                        &suggestion);
+              if (suggestion.empty()) {
+                return;
+              }
+
+              // Avoid showing the ending tag.
+              if (RE2::PartialMatch(suggestion, kResponseTagPattern)) {
+                return;
+              }
+
```
This code block is copy-pasted in ai_rewriter_ui.cc, please extract it into a util function in brave/components/ai_chat/core/browser/utils.h/cc to be shared.

petemill replied:
I had left it because I know we're going to remove it at some point - this shouldn't be done in this layer, it is engine-dependent and most engines do not need it since they strip it out with stop sequences. But I have made a util function in utils.cc - `BindParseRewriteReceivedData` which the rewriter dialog and in-line rewriter use. In the future we'll likely remove the in-line rewriter in favor of the rewriter dialog. I also removed all this from ConversationHandler as it does not need to be done here at all - nothing is calling it.

petemill commented on components/ai_chat/core/browser/ai_chat_service.cc:
```diff
@@ -0,0 +1,373 @@
+}
+
+size_t AIChatService::GetInMemoryConversationCountForTesting() {
+  CHECK(conversations_.size() == conversation_handlers_.size());
```
Well yes, without any DB then they should always be the same size, so this was me putting in a verification for that for the time-being during testing. But since this is only concerned with active conversations then I've changed it to return `conversation_handlers_.size()` and removed the CHECK - thanks.

petemill commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -0,0 +1,1385 @@
+
+  // Perf: make sure we're not doing this when the feature
+  // won't be used (e.g. no active conversation).
+  // DCHECK(is_conversation_active_)
+  //     << "UI shouldn't allow operations for an inactive conversation";
```
I don't think so, I think it's leftover from when this was being called automatically

petemill commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -0,0 +1,1385 @@
+void ConversationHandler::OnEngineCompletionDataReceived(
+    mojom::ConversationEntryEventPtr result) {
+  UpdateOrCreateLastAssistantEntry(std::move(result));
+}
```
I couldn't see why it was ever needed - the bool wasn't changed until the request is completed. We only need to call `OnAPIRequestProgressChanged` when that bool changes.

petemill commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -0,0 +1,1385 @@
+                       std::move(data_completed_callback), page_content,
+                       is_video));
+    return;
+  } else if (!should_refine_page_content && is_content_refined_) {
+    is_content_refined_ = false;
+    OnAssociatedContentInfoChanged();
+  }
```
If we shouldn't refine content, but we did refine content on the previous submission, then this property of SiteInfo has changed, so we notify the client via `OnAssociatedContentInfoChanged`.
It can happen if the page content got smaller or the model was changed.
I added a comment to make it clearer.

darkdh commented on components/ai_chat/core/browser/model_service_unittest.cc:
```diff
@@ -9,36 +9,167 @@
+  // Verify uses non-premium version
+  EXPECT_EQ(GetService()->GetDefaultModelKey(), "chat-claude-instant");
+  // Verify keeps non-premium version
+  GetService()->OnPremiumStatus(mojom::PremiumStatus::Inactive);
+  EXPECT_EQ(GetService()->GetDefaultModelKey(), "chat-claude-instant");
+  GetService()->OnPremiumStatus(mojom::PremiumStatus::Active);
+  EXPECT_EQ(GetService()->GetDefaultModelKey(), "chat-claude-instant");
```
shouldn't we also check `OnDefaultModelChanged`? same with test suite `MigrateOldClaudeDefaultModelKey`, `MigrateToPremiumDefaultModel` and `MigrateToPremiumDefaultModel_None` whichever calls `OnPremiumStatus`

petemill replied:
Thanks - added the event checks, and actually found a bug in the test and the tested code - models weren't being loaded before migration from `chat-claude-instant` pref value, so the intended replacements (chat-claude-haiku) weren't being found and the migration was ignored. The test case had the wrong value so was passing.

petemill commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -0,0 +1,1385 @@
+void ConversationHandler::MaybeUnlinkAssociatedContent() {
+  // Only unlink if panel is closed and there is no conversation history.
+  // When panel is open or has existing conversation, do not change the state.
+  // if (!is_conversation_active_ && chat_history_.empty()) {
+  if (chat_history_.empty()) {
```
I removed the comment. I don't see a reason this needs to delay the entry if there is no client connected. If `SubmitHumanConversationEntry` is called, it should be acted on.

darkdh replied:
I don't see changed reflected, maybe it is still in your local?

petemill replied:
I thought this was referring to another identical place! For this one it's right to add back in the check so I've done that now - thanks

petemill commented on components/ai_chat/core/browser/ai_chat_service.cc:
```diff
@@ -0,0 +1,373 @@
+  return conversation;
+}
+
+ConversationHandler* AIChatService::CreateConversationHandlerForContent(
```
de-duplicated with `MaybeAssociateContentWithConversation`

petemill commented on components/ai_chat/resources/page/components/alerts/long_conversation_info.tsx:
```diff
@@ -4,18 +4,17 @@
 
   const handleClearChat = () => {
-    getPageHandlerInstance().pageHandler.clearConversationHistory()
+    console.error('TODO: new conversation')
```
thanks - the ability to do thatcame later! I added the call to `aiChatContext.onNewConversation()`

petemill commented on components/ai_chat/resources/page/components/alerts/error_conversation_end.tsx:
```diff
@@ -21,7 +20,7 @@ function ErrorConversationEnd () {
           slot='actions'
           kind='plain-faint'
-          onClick={() => { getPageHandlerInstance().pageHandler.clearConversationHistory() }}
+          onClick={() => { console.error('TODO: new conversation') }}
```
added!

petemill commented on components/ai_chat/core/browser/conversation_handler.cc:
```diff
@@ -0,0 +1,1385 @@
+  }
+
+  CHECK(text_embedder_);
+  for (auto& callback_info : pending_top_similarity_requests_) {
```
done

petemill commented on browser/ui/ai_chat/ai_chat_tab_helper_unittest.cc:
```diff
@@ -0,0 +1,337 @@
+    EXPECT_EQ(helper_->GetPageURL(), url);
+  }
+
+  // void NavigateHistory(bool back = false,
```
Thanks, removed

darkdh commented on browser/ai_chat/ai_chat_service_factory.cc:
```diff
@@ -0,0 +1,85 @@
+    content::BrowserContext* context) {
+  CHECK(context);
+  if (!IsAllowedForContext(context) ||
+      !ModelServiceFactory::GetForBrowserContext(context) ||
+      !skus::SkusServiceFactory::GetForContext(context)) {
```
I think these can be safely removed because we have `DependsOn` to make sure those factories are accessible.

petemill replied:
as long as their requirements for context are the same

darkdh commented on browser/ai_chat/ai_chat_render_view_context_menu_browsertest.cc:
```diff
@@ -154,15 +122,42 @@ class AIChatRenderViewContextMenuBrowserTest : public InProcessBrowserTest {
+                    std::move(callback).Run(completed_result);
+                    run_loop.Quit();
+                  });
+          base::SingleThreadTaskRunner::GetCurrentDefault()->PostTask(
+              FROM_HERE, base::BindOnce(&ExecuteRewriteCommand, context_menu));
```
Why do we need to post executing rewrite command here?

petemill replied:
This is what the previous code did with `ContextMenuNotificationObserver` https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/renderer_context_menu/render_view_context_menu_browsertest_util.cc;l=37-40?q=ContextMenuNotificationObserver

Since that helper doesn't allow an additional intercepter for getting a handle to the `RenderViewContextMenu`, I had to replicate the basic functionality it does so that I could replace the EngineConsumer. I assume it has a good reason for posting to thread.

petemill commented:
Rebased on master, worked through re-implementing https://github.com/brave/brave-core/pull/25494 and https://github.com/brave/brave-core/pull/25513. Please take a look @darkdh @yrliou, especially at ai_chat_tab_helper.cc @darkdh because I did things a little differently. I don't think we need `pending_print_preview_requests_` and can get away with `pending_get_page_content_callback_` due to the simpler architecture in the refactor with print preview fallback. To avoid having to store navigation ID with the callback, I now reset it when `OnNewPage` is called.
I rebased the tests, but will have to run them tomorrow and see if I can add a couple more unit tests for AIChatTabHelper to simulate page loaded/not loaded.

darkdh commented on browser/ui/ai_chat/print_preview_extractor.cc:
```diff
@@ -0,0 +1,565 @@
+      SendResult("");
+      return;
+    }
+    memcpy(pdf_region.mapping.memory(), data->data(), data->size());
```
It is just code moving.

darkdh commented on components/ai_chat/content/browser/ai_chat_tab_helper.cc:
```diff
@@ -321,52 +278,105 @@ void AIChatTabHelper::GetPageContent(GetPageContentCallback callback,
+        std::move(pending_get_page_content_callback_).Run("", false, "");
+      }
+      pending_get_page_content_callback_ = std::move(callback);
+      return;
+    }
+    if (is_page_loaded_) {
+      // Fallback to print preview extraction
+      print_preview_extraction_delegate_->Extract(
+          IsPdf(web_contents()),
+          base::BindOnce(&AIChatTabHelper::OnExtractPrintPreviewContentComplete,
+                         weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+    }
```
else

petemill replied:
This wasn't quite an either/or but it's changed a lot now anyway

darkdh commented on components/ai_chat/content/browser/ai_chat_tab_helper.cc:
```diff
@@ -321,52 +278,105 @@ void AIChatTabHelper::GetPageContent(GetPageContentCallback callback,
+  if (base::Contains(kPrintPreviewRetrievalHosts, GetPageURL().host_piece()) &&
+      print_preview_extraction_delegate_ != nullptr) {
+    if (is_page_loaded_) {
+      // Get content using a printing / OCR mechanism, instead of
+      // directly from the source.
+      print_preview_extraction_delegate_->Extract(
+          is_pdf,
+          base::BindOnce(&AIChatTabHelper::OnExtractPrintPreviewContentComplete,
+                         weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
     } else {
-      FetchPageContent(web_contents(), invalidation_token, std::move(callback));
+      // Run print preview extraction after load
+      if (pending_get_page_content_callback_) {
+        std::move(pending_get_page_content_callback_).Run("", false, "");
+      }
+      pending_get_page_content_callback_ = std::move(callback);
```
I think this can be pull into something like `MaybePrintPreviewExact(...)` and use it here and `OnFetchPageContentComplete(...)`

petemill replied:
This is actually doing something slightly different to `OnFetchPageContentComplete` which is attempting to re-run regular `FetchPageContent` after page load, not just print preview extractor. So I've made 2 reusable functions to cover it.

darkdh reviewed (commented):
>  I don't think we need pending_print_preview_requests_ and can get away with pending_get_page_content_callback_ due to the simpler architecture in the refactor with print preview fallback

yes it can go away but it is mostly because the UI won't get refreshed every time we navigate away or reload in this refactor so users won't be able to create request callback belongs to different navigation ids.

petemill commented on browser/ai_chat/sources.gni:
```diff
@@ -4,6 +4,8 @@
 # You can obtain one at https://mozilla.org/MPL/2.0/.
 
 brave_browser_ai_chat_sources = [
+  "//brave/browser/ai_chat/ai_chat_service_factory.cc",
+  "//brave/browser/ai_chat/ai_chat_service_factory.h",
```
I just tried this, but the problem is:
```
> gn check /Users/petemill/Development/Brave/brave-browser/src/out/Component_arm64
ERROR Dependency cycle:
  //chrome/browser:browser ->
  //brave/browser/ai_chat:ai_chat ->
  //chrome/browser:browser
```

And that dependency would be needed because:
- We haven't moved brave/browser/skus and brave/browser/misc_metrics to their own targets
- The browser needs the AIChatServiceFactory

And if I try to remove the dependency from brave/browser and only put it on brave/browser/ui I still get a gn_check error:
```
ERROR Dependency cycle:
  //chrome/browser:browser ->
  //brave/browser/brave_wallet:brave_wallet_delegate ->
  //chrome/browser/ui:ui ->
  //brave/browser/ui:ui ->
  //brave/browser/ai_chat:ai_chat ->
  //chrome/browser:browser
```

This seems like something that's worth untangling as a whole over the whole codebase and in a separate PR?

yrliou replied:
Hm, yea, that's not ideal, it should only be //chrome/browser -> //brave/browser/ai_chat and not the other way around. That's an exact example why being in browser target itself would complicate things. :(

bridiver replied:
I'm not seeing why AIChatServiceFactory would need to depend on `//chrome/browser`. browser_process.h, brave_browser_process.h, profile_keyed_service_factory.h and profile.h are all in other targets specifically to fix these kinds of problems.

bridiver replied:
oic, you mentioned misc_metrics. We definitely should not be adding to the problem here or it's just going to be that much harder to untangle cc @DJAndries to fix the issue with misc_metrics. 
cc @bsclifton because SkusServiceFactory does not have any deps that would require it to be in sources.gni

bridiver replied:
I guess this PR was opened before we added codeowner for sources.gni because we're using that to get this issue under control

bridiver replied:
These both look like easy fixes, I don't see any reason why they needed to be in sources.gni in the first place. If there are any problems just split out interface/impl and put the header in its own target

petemill commented:
Hi reviewers - it's building and working on desktop, android and ios, and all tests / CI are passing (famous last words after another push!). I've addressed all feedback, apart from 2 things (from @darkdh) and was wondering whether we can defer for follow-up in the interest of getting it in master since they are nits:
- Adding more params to the ConversationHandler unit tests instead of separate tests with variables that I have now. I would like to do this in a follow-up because it would require rewriting 1300 lines of tests and doesn't very much change the coverage of the tests. We do in general need a lot more ConversationHandler tests, same as current master with ConversationDriver, but this PR improves it already.
- Putting common test functionality in a test_support / test_util file - this will take some time and doesn't improve coverage. It would be nice to remove duplication but there are negatives too.

petemill commented:
Everything passed CI except iOS due to unrelated test runner fix in https://github.com/brave/brave-core/pull/25669. Rebased to get that passing.

darkdh commented on components/ai_chat/core/browser/ai_chat_service_unittest.cc:
```diff
@@ -0,0 +1,485 @@
+      conversation_observations_{this};
+};
+
+class MockAIChatCredentialManager : public AIChatCredentialManager {
```
As discussed, we will do this in followup.

darkdh commented on components/ai_chat/core/browser/conversation_handler_unittest.cc:
```diff
@@ -0,0 +1,1249 @@
+  bool has_associated_content_ = true;
+};
+
+class ConversationHandlerUnitTest_OptedOut
+    : public ConversationHandlerUnitTest {
+ public:
+  ConversationHandlerUnitTest_OptedOut() { is_opted_in_ = false; }
+};
+
+class ConversationHandlerUnitTest_NoAssociatedContent
+    : public ConversationHandlerUnitTest {
+ public:
+  ConversationHandlerUnitTest_NoAssociatedContent() {
+    has_associated_content_ = false;
+  }
+};
```
As discussed, we will do this in followup.

bridiver commented on chromium_src/chrome/browser/autocomplete/chrome_autocomplete_provider_client.cc:
```diff
@@ -60,10 +62,16 @@ void ChromeAutocompleteProviderClient::OpenLeo(const std::u16string& query) {
   DCHECK(chat_tab_helper);
 
+  auto* conversation_handler =
+      ai_chat::AIChatServiceFactory::GetForBrowserContext(profile_)
```
don't we need a null check here before calling `GetOrCreateConversationHandlerForContent`?

petemill replied:
It was covered by
```
  CHECK(profile_->IsRegularProfile());
```
and the feature not being added to the omnibox suggestions in incognito profiles, but I agree it's better to check the service - added in 5f890855da16e83c3b3cf60a4b60937b88ed17a3

bridiver commented on chromium_src/chrome/browser/renderer_context_menu/render_view_context_menu.cc:
```diff
@@ -554,39 +554,58 @@ void BraveRenderViewContextMenu::ExecuteAIChatCommand(int command) {
+  } else {
+    auto* browser = GetBrowser();
+    if (!browser) {
+      VLOG(1) << "Can't get browser";
```
this kind of logging should be removed before merge

petemill replied:
this is not new

darkdh reviewed (approved):
ai-chat cpp side looks good to me. Unit tests for new codes and moved codes are nicely done with great coverage!

bridiver commented on chromium_src/chrome/browser/renderer_context_menu/render_view_context_menu.cc:
```diff
@@ -554,39 +554,58 @@ void BraveRenderViewContextMenu::ExecuteAIChatCommand(int command) {
+  auto selected_text = base::UTF16ToUTF8(params_.selection_text);
+
+  if (rewrite_in_place) {
+    source_web_contents_->SetUserData(kAIChatRewriteDataKey,
```
seems like this would be better handled inside the tab helper so it gets cleared when the page navigates? I think we have way too much logic for aichat embedded in RenderViewContextMenu here that should be encapsulated within aichat

bridiver replied:
also are there tests for this somewhere? I'm not seeing anything, but searching github PRs sucks so I could be missing something

bridiver replied:
nvm, I found them

petemill replied:
Sure but this is not added in this PR cc @yrliou

bridiver replied:
can we create follow-up issues to track this stuff?

petemill replied:
https://github.com/brave/brave-browser/issues/41171

brave-builds commented:
A Storybook has been deployed to [preview UI for the latest push](https://build-artifacts.brave.com/brave-browser-build-pr/aichat-services/storybook.html)

yrliou reviewed (approved):
My only outstanding comment is the issue of putting ai_chat_service_factory in browser target, I'll do an immediate follow-up to fix it.

petemill commented:
I made this issue to track findings of future code improvements https://github.com/brave/brave-browser/issues/41171
